stone/frontend/ir_generator.py

```diff
@@ -1248,3 +1259,235 @@ def mk_route_schema():
+            #     route_data_types.extend(namespace.get_route_io_data_types())
+            # else:
+            assert '*' not in routes_names
+            for route_name in routes_names:
+                if route_name not in namespace.route_by_name:
+                    print('Route %s is not defined!' % route_name, file=sys.stderr)
+                    sys.exit(1)
+                route = namespace.route_by_name[route_name]
+                route_data_types.extend(namespace.get_route_io_data_types_for_route(route))
+                if route.doc is not None:
+                    data_types, routes_by_ns = self._get_data_types_and_routes_from_doc_ref(
+                        route.doc, namespace_name)
+                    for d in data_types:
+                        route_data_types.append(d)
+                    for ns_name, routes in routes_by_ns.items():
+                        ns = self.api.namespaces[ns_name]
+                        for r in routes:
+                            for d in ns.get_route_io_data_types_for_route(r):
+                                route_data_types.append(d)
+
+        # Parse the datatype whitelist and populate any starting data types
+        for namespace_name, datatype_names in self._routes['datatype_whitelist'].items():
+            if namespace_name not in self.api.namespaces:
+                print('Namespace %s is not defined!' % namespace_name, file=sys.stderr)
+                sys.exit(1)
+
+            # Parse namespace doc refs and add them to the starting data types
+            namespace = self.api.namespaces[namespace_name]
+            if namespace.doc is not None:
+                data_types, routes_by_ns = self._get_data_types_and_routes_from_doc_ref(
+                    namespace.doc, namespace.name)
+                for d in data_types:
+                    route_data_types.append(d)
+                for ns_name, routes in routes_by_ns.items():
+                    ns = self.api.namespaces[ns_name]
+                    for r in routes:
+                        for d in ns.get_route_io_data_types_for_route(r):
+                            route_data_types.append(d)
+
+            for datatype_name in datatype_names:
+                if datatype_name not in self.api.namespaces[namespace_name].data_type_by_name:
+                    print('Namespace %s is not defined!' % namespace_name, file=sys.stderr)
+                    sys.exit(1)
+                data_type = self.api.namespaces[namespace_name].data_type_by_name[datatype_name]
+                route_data_types.append(data_type)
+
+        # Recurse on dependencies
+        output_types_by_ns, output_routes_by_ns = self._find_dependencies(route_data_types)
+
+        # Update the IR representation. This involves editing the data types and
+        # routes for each namespace.
+        for namespace in self.api.namespaces.values():
+            data_types = list(set(output_types_by_ns[namespace.name]))  # defaults to empty list
```

Comment:
FYI, technically sets() are also iterable.  Only reason you would need to convert a set() back to a list() is if you either need the resulting de-duplicated data accessible via index.

Reply:
yep, backend code (and mypy annotations) expect this to be a list, and uses some list-only functions too.